Project State

solnsumei opened the pull request:
### What does this PR do?
This PR shows the current state of the project
### Description
- Server side code base almost complete
- Client side react components added
- Server side test up to 80%
- Client side functionality almost complete

solnsumei committed:
Merge pull request #1 from solnsumei/create-middlewares-for-routes

Create middlewares for routes

solnsumei committed:
Add update method to booksController

solnsumei committed:
Finish update method in booksController

Write validation rules for query parameter and sanitize the bookId parameter. Test using postman for functionality.

solnsumei committed:
Merge pull request #2 from solnsumei/edit-books

Edit books

solnsumei committed:
Add model association to UserBook model

solnsumei committed:
Add borrow book method to usersController

solnsumei committed:
Fix bug in create-user-book migration

solnsumei committed:
Update borrow book method in usersController

Update method to restrict users from borrowing a book that is unavailable.
Write validation to prevent users from borrowing a book they have not returned twice.
Fix bug in UserBook model.

solnsumei committed:
Merge pull request #3 from solnsumei/borrow-book

Borrow book

solnsumei committed:
Fix eslint errors in files

Fix eslint errors in several files.
Install cross-env package to allow NODE_ENV on cross platforms

solnsumei committed:
Fix some more eslint errors in files

solnsumei committed:
Fix bug in the login method

solnsumei committed:
Finish API route for user to view borrowed books

Finish route to query borrow history and books not returned.
Write middleware to check if post requests contains a body object.
Write middleware to check query parameter for userId and also compare if userId passed is equal to the authenticated user.

solnsumei committed:
Add return book method in usersController

Clean up code base and add middleware to check book availability.

solnsumei committed:
Add houndCI configuration and .eslintignore files

solnsumei committed:
Refactor file names and add more test cases to user test

solnsumei committed:
Move test files to tests/test folder

solnsumei committed:
Add config files for testing and Heroku test deployment

solnsumei committed:
Edit server config.json

solnsumei committed:
Edit package.json file

solnsumei committed:
Edit server config files

solnsumei committed:
Migrate production test database

solnsumei committed:
Merge pull request #5 from solnsumei/view-borrowed-book

View borrowed book merge with development

solnsumei committed:
Rename config.js file to config.json

solnsumei committed:
Rename config.js file to config.json

solnsumei committed:
Fix bug in routes test cases

solnsumei committed:
Fix bug in book routes test

solnsumei committed:
Fix bug in production migration

solnsumei committed:
Fix bugs and added change middleware names

solnsumei committed:
Edit readMe.md and push app to origin

solnsumei committed:
Updated environment to test

solnsumei committed:
Write more API Endpoints docs to README file

solnsumei committed:
Edit README.md heading to reflect bold tag

solnsumei committed:
Update README for final change

solnsumei committed:
Complete README documentation

solnsumei committed:
Implement add quantity and delete methods for books

solnsumei committed:
Add book category model and migration table

solnsumei committed:
Add membership type model and migration table

solnsumei committed:
Refactor database schema

solnsumei committed:
Add categories controller functionality

solnsumei committed:
Add middleware for category controller

solnsumei committed:
Write method for user profile update

Write controller method for user profile update. Add middleware to check for valid membership types.

solnsumei committed:
Modify user update profile method

solnsumei committed:
Add the logout and change password method to user controller

solnsumei committed:
Modify JWT token implementation

solnsumei committed:
Commit changes for new pull request

solnsumei committed:
Add migration and seeder files

solnsumei committed:
Merge branch 'development' into test-deploy

solnsumei committed:
Merge pull request #9 from solnsumei/test-deploy

Pull request from test-deploy branch

solnsumei committed:
Add more dependencies to package.json

solnsumei committed:
Update package.json file

solnsumei committed:
Update gitignore to exclude migrations and seeders folders

solnsumei committed:
Fix eslint errors in books.js

solnsumei committed:
Correct eslint errors in users.js and categories.js files

solnsumei committed:
Correct eslint errors in middleware files

solnsumei committed:
Update config.json to use new db credentials

solnsumei committed:
Correct eslint issues in model files

solnsumei committed:
Modify config.json to enable database migration

solnsumei committed:
Merge pull request #10 from solnsumei/notifications

Notifications

solnsumei committed:
Create src folder to hold source client files

solnsumei committed:
Set up webpack test files

solnsumei committed:
Run test jsx component

solnsumei committed:
Modify webpack to compile css files

solnsumei committed:
Add test component and compile

solnsumei committed:
Merge pull request #11 from solnsumei/setup-react

Commit setup-react branch to development branch

solnsumei committed:
Save server config file for webpack

solnsumei committed:
Create register component class

solnsumei committed:
Create catalog react component

solnsumei committed:
Modify the book detail component

solnsumei committed:
Add the book component

solnsumei committed:
Modify Book detail component to use book component

solnsumei committed:
Add book history and borrowed item components

solnsumei committed:
Add borrow history component

solnsumei committed:
Add profile component

solnsumei committed:
Add top title component

solnsumei committed:
Finish webpack config for scss to css

solnsumei committed:
Edit read me and add badges

solnsumei committed:
Merge pull request #13 from solnsumei/create-components

Create components

solnsumei committed:
Update readme

solnsumei committed:
Merge pull request #14 from solnsumei/create-components

Update readme

solnsumei committed:
Modify readMe

solnsumei committed:
Merge pull request #15 from solnsumei/create-components

Modify readMe

solnsumei committed:
Update readme

solnsumei committed:
Merge pull request #16 from solnsumei/create-components

Update readme

solnsumei committed:
Fix bug in package.json

solnsumei committed:
Update travis.yml and package.json with test config

solnsumei committed:
Update travis to fix bugs

solnsumei committed:
Fix bugs in travis.yml file

solnsumei committed:
Add migration and seeder files to git

solnsumei committed:
Edit travis.yml and package.json files

solnsumei committed:
Update package.json to fix bug in test

solnsumei committed:
Update travis.yml to set node environment

solnsumei committed:
Update scripts to create test database

solnsumei committed:
Update script to support tests

solnsumei committed:
Add cross-env package to save environment

solnsumei committed:
Update repo tokens

solnsumei committed:
Add codeclimate test reporter id

solnsumei committed:
Merge pull request #27 from solnsumei/create-components

Create components

solnsumei committed:
Edit tests to support new database schema

solnsumei committed:
Add more test codes

solnsumei committed:
Fix bug in user test

solnsumei committed:
Fix comma bug in test

solnsumei committed:
Fix bug in book test

solnsumei committed:
Ignore admin seeder from test data

solnsumei committed:
Fix bug in category test data

solnsumei committed:
Fix bugs in tests

solnsumei committed:
Modify return messages to suit test output

solnsumei committed:
Fix bug in test code

solnsumei committed:
Fix bug in book routes test

solnsumei committed:
Merge pull request #29 from solnsumei/create-components

Merge create components to development

solnsumei committed:
Fix issues with test coverage reporting

solnsumei committed:
Merge pull request #30 from solnsumei/create-components

Fix issues with test coverage reporting

solnsumei committed:
Add codeclimate config file

solnsumei committed:
Merge pull request #31 from solnsumei/create-components

Add codeclimate config file

solnsumei committed:
Update files and file path

solnsumei committed:
Fix bug in profile page

solnsumei committed:
Merge pull request #32 from solnsumei/create-components

Create components

solnsumei committed:
Add boilerplate for signup form

solnsumei committed:
Add demo config files for redux actions

solnsumei committed:
Add cross origin permission for client connection to server

solnsumei committed:
Merge pull request #33 from solnsumei/add-signup-page

Add signup page

solnsumei committed:
Update package to fix test coverage config

solnsumei committed:
Merge pull request #34 from solnsumei/add-signup-page

Update package to fix test coverage config

solnsumei committed:
Add dependencies for test coverage reporting

solnsumei committed:
Merge pull request #35 from solnsumei/add-signup-page

Add dependencies for test coverage reporting

solnsumei committed:
feature(update): Implement sig up functionality from client side

solnsumei committed:
Merge pull request #36 from solnsumei/add-signup-page

feature(update): Implement sign up functionality from client side

solnsumei committed:
feature(user-login): Create login Form and Login Page components

solnsumei committed:
feature(modify): Update signUp implementation to use history

solnsumei committed:
feature(login-failed): Added login failure notification message

solnsumei committed:
feature(add-auth): Add auth helper function to set tokens

solnsumei committed:
feature(maintain): Fix server crash error and modify login

solnsumei committed:
feature(login-success-action): Update user reducers to update user state

solnsumei committed:
feature(add-redrirects): Add redirect actions and reducers

solnsumei committed:
feature(modify): Modify redirect reducer to return appropriate redirect url

solnsumei committed:
feature(modify): Finish implementation of authentication redirects

solnsumei committed:
feature(logout-user): Implement logout feature

solnsumei committed:
Merge pull request #37 from solnsumei/add-login

Add login pull request

solnsumei committed:
feature(show-profile-details): Add profile view functionality

solnsumei committed:
faeture(update): Remove page title actions

solnsumei committed:
feature(sign-up-redirect): Edit signup action to login user also

solnsumei committed:
feature(add-membership-type): Add membership type controller

solnsumei committed:
feature(add-membershipttype-middleware): Add Membership Type Middleware

solnsumei committed:
feature(update-server-route): Add middleware to membership update request

solnsumei committed:
feature(add-edit-profile-form): Add Edit Profile Form

solnsumei committed:
feature(modify): Modify update user method in server controller

solnsumei committed:
feature(modify): Add global scope for jquery initialization

solnsumei committed:
feature(update-user-profile): Implement update profile functionality

solnsumei committed:
Merge pull request #38 from solnsumei/user-profile

User profile

solnsumei committed:
feature(middleware): Add admin middleware component

solnsumei committed:
feature(admin-dashboard): Add admin dashboard

solnsumei committed:
feature(add-category): Add category functionality

solnsumei committed:
feature(bug-fix): Fix log out issues with state management

solnsumei committed:
feature(modify): Create app reducer to handle initialization of store

solnsumei committed:
feature(add-quantity): Add AddQuantity modal and refactor constants

solnsumei committed:
feat(delete-book): delete book functionality started

solnsumei committed:
feat(maintain): update users controller borrowBook and returnBook method

solnsumei committed:
feat(borrowed-books): add borrowed books not returned functionality

solnsumei committed:
feat(add-test): Add more test cases

solnsumei committed:
feat(maintain): Fix bug in category test

solnsumei committed:
feat(maintain): Update category test cases

solnsumei committed:
feat(maintain): Update test case abd dataholder

solnsumei committed:
bug(fix): Fix bug in category test

solnsumei committed:
feat(maintain): update test case

solnsumei committed:
Merge pull request #39 from solnsumei/admin-functions

Admin functions

solnsumei committed:
feat(update-test): Add membership test case

solnsumei committed:
Fix(bug): Fix bugs in membershiptype test case

solnsumei committed:
fix(membership-type): Fix test case to send proper response

solnsumei committed:
feat(update-test): Add more test cases to books

solnsumei committed:
feat(add-test): Fix issue with default category in book routes test

solnsumei committed:
update test

solnsumei committed:
Update test cases

solnsumei committed:
update book tests to reflect borrow book

solnsumei committed:
Update test parameters for borrow and return books

solnsumei committed:
Update eslint file to fix errors

solnsumei committed:
Merge pull request #40 from solnsumei/admin-functions

Merge update admin actions

solnsumei committed:
feat(maintain): update code base to reflect code review feedback

solnsumei committed:
feat(maintain): Rename membershipTypes controller to membership

solnsumei committed:
feat(maintain): Update package.json file for build configuration

solnsumei committed:
feat(update): modify file case

solnsumei committed:
fix(bug): Commit file name chages

solnsumei committed:
Fix(bug): Rename files to effect git change

solnsumei committed:
fix(bug): Fix file name issues

solnsumei committed:
fix(bug): fix git issues with file renaming

solnsumei committed:
fix(bug): fix bug in middlewares error checking

solnsumei committed:
feat(refactor): Refactor code with DRY principle

solnsumei committed:
feat(maintain): remove commented out code

solnsumei committed:
feat(maintain): add admin functions

- Rearrange components into respective folders
- Ignore .env files from git commits
- Rewrite ReadME.md file

solnsumei committed:
feat(maintain): ignore .env file

solnsumei committed:
feat(maintain): remove console.log from editMembershipTypeRequest.js

solnsumei committed:
Commit

solnsumei committed:
fix(bug): fix bugs in controller and test cases

solnsumei committed:
Merge pull request #42 from solnsumei/admin-functions

Admin functions

solnsumei committed:
feat(change-password): Add controller and middleware method

- Create middleware changePasswordRequest
- Add changePassword method to API users controller
- Create users/change-password route

solnsumei committed:
feat(maintain): update project interface

solnsumei committed:
feat(change-password): Add change password functionality

- UI redesign
- add change password functionality on server side
- add change password page
- add change password form

[Finishes #149783557]

solnsumei committed:
fix(bug): Update package.json file to reflect production dependencies

solnsumei committed:
fix(bug-fix): Fix bug in heroku post build

solnsumei committed:
fix(bug-fix): Correct code in package.json to copy index.html to dist

solnsumei committed:
fix(bug-fix): Fix https issues with cloudinary widget

solnsumei committed:
Merge pull request #44 from solnsumei/feature/149783557/change-password

Feature/149783557/change password

solnsumei committed:
feat(add-test-case): add more test cases for routes

solnsumei committed:
fix(bug-fix): fix bug in omitted import statement for test case

solnsumei committed:
fix(bug-fix): fix bug with parsing null values to test

solnsumei committed:
fix(bug-fix) fix bug in passing wrong url in category test route

solnsumei committed:
fix(bug-fix): remove unused function in categoryRequest middleware

solnsumei committed:
chore(add-tests): update category route test

solnsumei committed:
refactor(components):  Rename all components with the jsx extension

- rename react components
- update webpack config
- update babelrc  file
- update eslintrc file

jetanuwoma commented on client/src/app/actions/bookActions.js:
```diff
@@ -31,6 +35,14 @@ const loadBooks = headers => dispatch =>
+  axios.get(`/api/v1/books/${bookId}`, headers)
+    .then(({ data }) => dispatch(getBookSuccess(data.book)))
+    .catch((error) => {
+      throw (error);
```
Don't just throw any error, try to customise it so your users can know what's wrong

jetanuwoma commented on client/src/app/actions/userActions.js:
```diff
@@ -68,6 +68,11 @@ const updateUserAccount = userData => (dispatch) => {
+const changeUserPassword = passwordObject => (dispatch) => {
+  const headers = authCheck(dispatch);
+
+  return axios.post('/api/v1/users/change-password', passwordObject, headers);
```
Try to abstract `api/v1` and save as axios default route

jetanuwoma commented on client/src/app/components/admin/books/BooksPage.js:
```diff
@@ -37,6 +37,15 @@ class BooksPage extends React.Component {
   /**
+   * [componentDidMount description]
+   * @method componentDidMount
+   * @return {[type]}          [description]
```
Supply a valid return type and param type

jetanuwoma commented on client/src/app/components/book/BookDetailPage.js:
```diff
@@ -4,16 +4,17 @@ import toastr from 'toastr';
    * @method constructor
-   * @param  {[type]}    props [description]
-   * @return {[type]}          [description]
+   * @param  {obj} props react properties
```
Change this to `Object`

solnsumei committed:
docs(README.md): Update README.md file

- Update README.md file
- Add a LICENSE.md file

solnsumei committed:
docs(update-readme): Update README.md file to reflect doc changes

solnsumei committed:
refactor(add-docs): update docs in components

- update package.json to ignore seeding adminData to test db
- ignore nyc_output and coverage reports from git
- update README.md file
- update webpack_config to compile both css files and jsx files
-  refactor some server functions

enahomurphy commented on client/src/app/components/admin/books/BooksPage.js:
```diff
@@ -0,0 +1,197 @@
+    this.props.deleteBook(this.state.book)
+      .then(() => {
+        $('.modal').modal('close');
+        this.setState({ book: Object.assign({}, {}) });
```
Why not just assign the property a new object 

enahomurphy commented on client/src/app/components/common/LoadingDots.js:
```diff
@@ -0,0 +1,59 @@
+  render() {
+    let dots = this.state.frame % (this.props.dots + 1);
+    let text = '';
+    while (dots > 0) {
```
What is this for

enahomurphy commented on server/controllers/books.js:
```diff
@@ -0,0 +1,217 @@
+          }
+        })
+          .catch((error) => {
+            if (error.name === 'SequelizeValidationError' ||
```
The same issue with multiple code repetition., I can see the use of this code in multiple places

enahomurphy commented on server/controllers/users.js:
```diff
@@ -0,0 +1,308 @@
+    return db.User
+      .findById(req.auth.id)
+      .then((user) => {
+        if (user) {
```
this if and the one below can be evaluated on one line

enahomurphy commented on server/controllers/users.js:
```diff
@@ -0,0 +1,308 @@
+              .catch((error) => {
+                if (error.name === 'SequelizeValidationError') {
+                  const errors = {};
+                  error.errors.forEach((err) => {
```
I can see multiple uses of the loop, this can be extracted into a function

enahomurphy commented on server/middlewares/categoryRequest.js:
```diff
@@ -0,0 +1,88 @@
+ * @returns {Request|Response|*|void|boolean} res
+ */
+export function validateCategoryId(req, res, next) {
+  if (req.body.categoryId === undefined || req.body.categoryId === null ||
```
is there a need for this if you have a check using parseInt

refinedblessing replied:
Seeing you are using req.body.categoryId a lot, just assign to a variable instead at the start of the function

refinedblessing commented on server/controllers/books.js:
```diff
@@ -0,0 +1,217 @@
+import db from '../models/index';
+import { formatBookObject } from '../helpers/formatData';
+/**
+ * Controller for adding, updating and get all books
```
`get` here should be getting

refinedblessing commented on server/controllers/users.js:
```diff
@@ -0,0 +1,308 @@
+                .then(result => res.status(200).send({ message: 'Book was returned successfully',
+                  returnedBook: formatBorrowedBookObject(borrowedBook, req.book),
+                }))
+                .catch(error => res.status(500).send({
```
This line and some other lines are basically doing the same thing, can you have a separate error handler for this.
http://expressjs.com/en/guide/error-handling.html

refinedblessing commented on server/middlewares/categoryRequest.js:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @returns {Request|Response|*|void|boolean} res
+ */
+export function validateCategoryIdParam(req, res, next) {
```
validateCategoryIdParam and validateCategoryId can be put together, you can check for the categoryId in req.body or req.params and do the same thing. Also, use only the last check in the if statement.

refinedblessing commented on server/middlewares/categoryRequest.js:
```diff
@@ -0,0 +1,88 @@
+ * @returns {Request|Response|*|void|boolean} res
+ */
+function validateCategory(req, res, next, id) {
+  return db.Category.findById(id)
```
I think you can use the findCategory method here instead of making another request to the db.

refinedblessing commented on server/middlewares/changePasswordRequest.js:
```diff
@@ -0,0 +1,37 @@
+export default function checkPasswordRequest(req, res, next) {
+  const errors = {};
+
+  if (req.body.oldPassword === undefined || req.body.oldPassword === null) {
```
Can you have a function that checks if the value passed to it is null or undefined instead of doing a similar check in the same long way?

refinedblessing replied:
You do this in a lot of the files, so you might want to extract it to a util file.

refinedblessing commented on server/middlewares/editMembershipTypeRequest.js:
```diff
@@ -0,0 +1,41 @@
+  const errors = {};
+
+  if (req.params.membershipTypeId === undefined || req.params.membershipTypeId === null ||
+    !Number.isInteger(parseInt(req.params.membershipTypeId, 10))) {
```
Only the last check here will do

solnsumei committed:
refactor(app-code): apply code review feedback

- refactor code in controllers to remove duplicate codes
- refactor code in middleware
- refactor react components
- write more tests
- create shared error handler
- add validation package

solnsumei committed:
bug(fix-bug): fix bug in travis.yml file

solnsumei committed:
bug(fix-bug): modify axios default header

solnsumei committed:
refactor(app): implement review

- update controller methods
- update database method
- refactor code in helpers and middlewares
- refactor code in client side components

solnsumei committed:
feat(admin-notification): add client side notification

- add jsx files for notification page, item and modal
- write actions and reducers for notifications
- add code for initialization
- read notification functionality

[Finishes #149783500]

solnsumei committed:
Merge pull request #45 from solnsumei/feature/149783500/admin-notification-for-borrowing-activities

#149783500 Admin Notification for Borrowing Activities

solnsumei committed:
feat(upgrade-user_level): upgrade user level

- refactor middlewares
- update authentication check
- refactor route files
- add controller code to upgrade user level

[Finishes #153006340]

solnsumei committed:
fix(bug-fix): fix bug in loading borrowed books on login

solnsumei committed:
Merge pull request #46 from solnsumei/feature/153006340/upgrade-user-membership-level

#153006340 Upgrade User Membership Level

solnsumei committed:
feature(pagination): add pagination

- write server side pagination function
- add pagination to controllers
- create a pagination stateless component
- add pagination to books, categories, and notifications actions
- add pagination component to pages

[Finishes #153118111]

solnsumei committed:
Merge pull request #47 from solnsumei/feature/153118111/pagination

#153118111 Pagination

solnsumei committed:
feat(google-login): implement google sign up

- install react-google-login package
- implement google login on client-side
- add method to check google user on the server side

[Finishes #149783603]

solnsumei committed:
Merge pull request #48 from solnsumei/feature/149783603/google-signup

#149783603 Google Signup

solnsumei committed:
feat(reset-password): reset pasword

-

solnsumei committed:
feat(reset-password): reset password

- install nodemailer package
- setup account with sendgrid
- create password controller
- add client side components for forgot password page
- add client side components for reset password
- add server side implementation for reset and forgot password

[Finishes #149783618]

solnsumei committed:
Merge pull request #49 from solnsumei/feature/149783618/reset-password

#149783618 Reset Password

solnsumei committed:
modify mailHelper with environment credentials

solnsumei committed:
refactor(mailHelper): refactor mail helper

solnsumei committed:
feat(surcharge): send surcharge for late  return

- install cron package
- write functions to check defaulters
- send mail to alert surcharged users

[Finishes #49783518]

solnsumei committed:
refactor(borrow-access):  add surcharge check

- add surcharge check to borrowAccess middleware

solnsumei committed:
Merge branch 'development' into feature/149783518/surcharge

solnsumei committed:
fix(bug-fix): fix bug in borrowAccess middleware

[Fixes #149783518]

solnsumei committed:
Merge pull request #50 from solnsumei/feature/149783518/surcharge

#149783518 Surcharge User For Late Book Return

solnsumei committed:
test(add-test): write test for app

- add client side action, reducer and components test
- update server side test for new fetures added
- modify app  files to pass test

[Delivers #153340130]

solnsumei committed:
test(update-test): update test to exclude nodemailer actions

solnsumei committed:
Merge pull request #51 from solnsumei/chore/153340130/testing

#153340130 Add Tests

solnsumei committed:
chore(api-docs) : add api documentation

solnsumei committed:
fix(ignore-docs): add api docs folder to eslintignore

solnsumei committed:
Merge pull request #52 from solnsumei/chore/153627516/api-docs

#153627516 Add API Documentation

solnsumei committed:
test(e2e-test): write end to end test

- update code base to implement jsdoc comment
- add end to end test

solnsumei committed:
fix(bug-fix): fix bug in functions test

solnsumei committed:
docs(jsdoc): add jsdoc comments

solnsumei committed:
refactor(refactor-code): refactor code to enable server files build

- add jsdoc comments to react components
- add webpack cofiguration for production
- move app.js from root folder to server folder
- change constants.js helper filename to routeTitle.js

solnsumei committed:
refactor(refactor-code): refactor code to enable server files build

    - add jsdoc comments to react components
    - add webpack cofiguration for production
    - move app.js from root folder to server folder
    - change constants.js helper filename to routeTitle.js

solnsumei committed:
doc(add-jsdoc): add jsdoc comments to functions and classes

- add jsdoc comments to files
- update README.md documentation
- added a .env.example file

solnsumei committed:
fix(bug-fix): correct test path for moved app.js file

solnsumei committed:
fix(bug-fix): correct app.js path for server-side test files

solnsumei committed:
fix(bug-fix): fix bug in template string

solnsumei committed:
Merge pull request #53 from solnsumei/chore/153625894/e2e-test

#153625894 E2E Test
